pkg/webhook/handlers.go

```diff
@@ -0,0 +1,165 @@
+	// Create the handler
+	h := &resourceHandler{}
+	found := false
+	if v, ok := o.(Validator); ok {
+		h.Validator = v
+		found = true
+	}
+	if d, ok := o.(Defaulter); ok {
+		h.Defaulter = d
+		found = true
+	}
+	if !found {
+		// Type doesn't implement functions for webhooks
+		return nil, nil
+	}
+
+	// Create the webhook
+	return builder.NewWebhookBuilder().
+		Mutating(). // Always use a mutating webhook in case defaulting is implemented later
+		Operations(
+			admissionregistrationv1beta1.Create,
+			admissionregistrationv1beta1.Update).
+		WithRestMapper(restMapper).
+		WithScheme(scheme).
+		ForType(o).
+		Handlers(h).
+		Build()
+}
+
+type resourceHandler struct {
+	Validator Validator
+	Defaulter Defaulter
+	Decoder   types.Decoder
+}
+
+var _ inject.Decoder = &resourceHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *resourceHandler) InjectDecoder(d types.Decoder) error {
+	h.Decoder = d
+	return nil
+}
+
+// Handle handles admission requests.
+func (h *resourceHandler) Handle(ctx context.Context, req types.Request) types.Response {
+	// Get the object in the request
+	if resp, send := h.doValidate(ctx, req); send {
+		return resp
+	}
+
+	if resp, send := h.doDefault(ctx, req); send {
+		return resp
+	}
+
+	return admission.ValidationResponse(true, "")
+}
+
+func (h *resourceHandler) doDefault(ctx context.Context, req types.Request) (types.Response, bool) {
+	if h.Defaulter == nil {
+		return types.Response{}, false
+	}
+
+	// Get the object in the request
+	obj := h.Defaulter.DeepCopyObject().(Defaulter)
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err), true
+	}
+
+	// Default the object
+	copy := obj.DeepCopyObject().(Defaulter)
+	copy.Default()
+
+	// Create the patch
+	return admission.PatchResponse(obj, copy), true
+}
+
+func (h *resourceHandler) doValidate(ctx context.Context, req types.Request) (types.Response, bool) {
+	if h.Validator == nil {
+		return types.Response{}, false
+	}
+
+	// Get the object in the request
+	obj := h.Validator.DeepCopyObject().(Validator)
+	if req.AdmissionRequest.Operation == v1beta1.Create {
+		err := h.Decoder.Decode(req, obj)
+		if err != nil {
+			return admission.ErrorResponse(http.StatusBadRequest, err), true
+		}
+
+		err = obj.ValidateCreate()
+		if err != nil {
+			return admission.ValidationResponse(false, err.Error()), true
+		}
+	}
+
+	if req.AdmissionRequest.Operation == v1beta1.Update {
+		oldObj := obj.DeepCopyObject()
+
+		err := h.Decoder.Decode(req, obj)
+		if err != nil {
+			return admission.ErrorResponse(http.StatusBadRequest, err), true
+		}
+		err = h.Decoder.DecodeOld(req, oldObj)
+		if err != nil {
+			return admission.ErrorResponse(http.StatusBadRequest, err), true
+		}
+
+		err = obj.ValidateUpdate(oldObj)
+		if err != nil {
+			return admission.ValidationResponse(false, err.Error()), true
+		}
+	}
+
+	return types.Response{}, false
+}
```

Comment:
This is fantastic. I like the way it hides all the webhook machinery from the end-users. Resource author just need to define the `validate` and `default` methods on their types and rest is taken care by the machinery (exactly the way it should be). 

This is a huge step in making `defaulting` and `validation` accessible.

Reply:
+1